src/cls/rbd/cls_rbd.cc

```diff
@@ -1795,6 +1898,169 @@ int dir_remove_image(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+ * @param default_state the default state of newly created objects
+ *
+ * Output:
+ * @returns 0 on success, negative error code on failure
+ */
+int object_map_resize(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  uint64_t object_count;
+  uint8_t default_state;
+  try {
+    bufferlist::iterator iter = in->begin();
+    ::decode(object_count, iter);
+    ::decode(default_state, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  BitVector<2> object_map;
+  int r = object_map_read(hctx, object_map);
+  if ((r < 0) && (r != -ENOENT)) {
+    return r;
+  }
+
+  size_t orig_object_map_size = object_map.size();
+  if (object_count < orig_object_map_size) {
+    for (uint64_t i = object_count; i < orig_object_map_size; ++i) {
+      if (object_map[i] != default_state) {
+	return -ESTALE;
+      }
+    }
+    object_map.resize(object_count);
+  } else if (object_count > orig_object_map_size) {
+    object_map.resize(object_count);
+    for (uint64_t i = orig_object_map_size; i < object_count; ++i) {
+      object_map[i] = default_state;
+    }
+  }
+
+  bufferlist map;
+  ::encode(object_map, map);
+  CLS_LOG(20, "object_map_resize: object size=%llu, byte size=%llu",
+	  static_cast<unsigned long long>(object_count),
+	  static_cast<unsigned long long>(map.length()));
+  return cls_cxx_write_full(hctx, &map);
+}
+
+/**
+ * Update an rbd image's object map
+ *
+ * Input:
+ * @param start_object_no the start object iterator
+ * @param end_object_no the end object iterator
+ * @param new_object_state the new object state
+ * @param current_object_state optional current object state filter
+ *
+ * Output:
+ * @returns 0 on success, negative error code on failure
+ */
+int object_map_update(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  uint64_t start_object_no;
+  uint64_t end_object_no;
+  uint8_t new_object_state;
+  boost::optional<uint8_t> current_object_state;
+  try {
+    bufferlist::iterator iter = in->begin();
+    ::decode(start_object_no, iter);
+    ::decode(end_object_no, iter);
+    ::decode(new_object_state, iter);
+    ::decode(current_object_state, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  BitVector<2> object_map;
+  int r = object_map_read(hctx, object_map);
+  if (r < 0) {
+    return r;
+  }
+
+  if (start_object_no >= end_object_no || end_object_no > object_map.size()) {
+    return -ERANGE;
+  }
+
+  bool updated = false;
+  for (uint64_t object_no = start_object_no; object_no < end_object_no;
+       ++object_no) {
+    if ((!current_object_state || object_map[object_no] == *current_object_state) &&
+	object_map[object_no] != new_object_state) {
+      object_map[object_no] = new_object_state;
+      updated = true;
+    }
+  }
+
+  if (updated) {
+    uint64_t byte_offset, byte_length;
+    object_map.get_data_extents(start_object_no,
+				end_object_no - start_object_no,
+				&byte_offset, &byte_length);
+
```

Comment:
Hmm, bitvector uses encode and decode which can't directly parse partial raw bytes.

I guess that objectmap may need to construct a new disk-format structure so we don't need to encode/decode bitvector, maybe like this:

Struct ObjectMap {
uint64_t flags; crc/feature bits
uint64_t size;
uint64_t data;
uint64_t magic_pad;
}; (packed)

It may be more flexible for caller?